### PR TITLE
Preserve original members of a union type and add support to check if a union corresponds to an enum at runtime

### LINKED ISSUE #28782: `toString()` of union types with a name should include the module
**Description:**
$title.

**Steps to reproduce:**
```ballerina
import ballerina/lang.value;

public function main() {
    value:Cloneable x = 1;
    any y = x;
}

type Cloneable record {|
    
|};
```

Currently the error is just
```cmd
$ bal run qwe.bal 

Compiling source
        qwe.bal
ERROR [qwe.bal:(5:13,5:14)] incompatible types: expected 'any', found 'Cloneable'
```
But this can be insufficient and misleading.

**Affected Versions:**
slalpha2

### LINKED ISSUE #29139: Need a way to Identify Enum Types at Runtime
**Description:**
For the Ballerina GraphQL package, we need to support `enum` types [ballerina-standard-library#1000](https://github.com/ballerina-platform/ballerina-standard-library/issues/1000).

Since we are generating the GraphQL schema in the runtime, we have to have a way to identify `enum` types in the runtime, in order to complete this. Currently, the `enum` type is identified as a union type, and there's no way to differentiate other union types from an enum.

### LINKED ISSUE #29654: Immutable cyclic union results in a bad, sad error at runtime
**Description:**
$title.

Note: currently the name of the immutable type seems to be the same as the original type for unions. https://github.com/ballerina-platform/ballerina-lang/pull/29362 fixes the name, but it still fails.

**Steps to reproduce:**
```ballerina
import ballerina/lang.value;

public function main() {
    value:Cloneable & readonly x = 1;
}
```

```log
[2021-03-27 00:31:42,874] SEVERE {b7a.log.crash} - $type$Cloneable 
java.lang.NoSuchFieldError: $type$Cloneable
	at foo.main(foo.bal:4)
	at $_init.$lambda$main$(.)
	at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:571)
	at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:301)
	at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:269)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Error with change in #29362.
```log
[2021-03-27 00:33:10,004] SEVERE {b7a.log.crash} - $type$(Cloneable & readonly) 
java.lang.NoSuchFieldError: $type$(Cloneable & readonly)
	at ballerina.lang$0046value.1_0_0.$_init.$populate$type$Cloneable(lang.value)
	at ballerina.lang$0046value.1_0_0.$_init.$createTypes(lang.value)
	at ballerina.lang$0046value.1_0_0.$_init.$currentModuleInit(lang.value)
	at ballerina.lang$0046value.1_0_0.$ConfigurationMapper.$initAndPopulateConfigData(Unknown Source)
	at $ConfigurationMapper.$configureInit(Unknown Source)
	at $_init.main(.)
```

**Affected Versions:**
Current master

### LINKED ISSUE #29656: Immutable cyclic union isn't immutable
**Description:**
$title. The name of the immutable union is currently the same as the original union. So a new type doesn't seem to get created.

The name is changed in https://github.com/ballerina-platform/ballerina-lang/pull/29362. Still fails for Cloneable as mentioned in #29654.

**Steps to reproduce:**
The following does not panic.
```ballerina
type MyCyclicUnion int|MyCyclicUnion[];

public function main() {
    MyCyclicUnion x = [1, 2];
    MyCyclicUnion & readonly y = <MyCyclicUnion & readonly> x;
}
```

**Affected Versions:**
Current master (SLAlpha4-SNAPSHOT)

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/29654

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

This PR also cleans up some BUnionType constructors.

Fixes #29139 - Need a way to Identify Enum Types at Runtime
Fixed #28782 - `toString()` of union types with a name should include the module

This PR also fixes issues with immutable union (cyclic) types. If an immutable union type is created for a mutable union type with a name, a new type def is added for the immutable union type too - similar to records and objects.

Fixes #29654 - Immutable cyclic union results in a bad, sad error at runtime
Fixes #29656 - Immutable cyclic union isn't immutable

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java [hunk 1]
```diff
@@ -502,4 +464,60 @@ public void mergeUnionType(BUnionType unionType) {
         }
         setFlagsBasedOnMembers();
     }
+
+    private String computeStringRepresentation() {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java:468]: This logic is now the same as that at compile-time.

**Comment by hasithaa** [comment]
@MaryamZi There is a local build failure with this PR with parallel merge. 

failing following, `./gradlew :ballerina-runtime:test`

output 
```
ballerina-runtime-test-suite > ballerina-runtime-test > io.ballerina.runtime.test.cli.OptionTest > testInvalidUnionType FAILED
    java.lang.AssertionError: expected [unsupported type expected with main function 'string|int'] but found [unsupported type expected with main function '(string|int)']
        at org.testng.Assert.fail(Assert.java:96)
        at org.testng.Assert.failNotEquals(Assert.java:776)
        at org.testng.Assert.assertEqualsImpl(Assert.java:137)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:453)
        at org.testng.Assert.assertEquals(Assert.java:463)
        at io.ballerina.runtime.test.cli.OptionTest.testInvalidUnionType(OptionTest.java:145)
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java [hunk 2]
```diff
@@ -114,49 +93,34 @@ public BUnionType(List<Type> memberTypes, boolean readonly) {
...
-     */
-    public BUnionType(int typeFlags, boolean readonly, boolean isCyclic) {
-        super(null, null, Object.class);
+    public BUnionType(String name, Module pkg, int typeFlags, boolean isCyclic, long flags) {
+        super(name, pkg, Object.class);
         this.typeFlags = typeFlags;
-        this.readonly = readonly;
-        this.memberTypes = new ArrayList<>(2);
+        this.readonly = isReadOnlyFlagOn(flags);
+        this.memberTypes = null;
```
Comment (grainier) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java:116]: Can't this lead to NPEs? Shouldn't we have an empty list instead?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java [hunk 3]
```diff
@@ -55,55 +60,29 @@
...
-    public BUnionType(List<Type> memberTypes, int typeFlags) {
-        this(memberTypes, typeFlags, false, false);
-    }
-
     public BUnionType(List<Type> memberTypes, int typeFlags, boolean readonly,  boolean isCyclic) {
         super(null, null, Object.class);
         this.typeFlags = typeFlags;
         this.readonly = readonly;
-        setMemberTypes(memberTypes);
+        this.flags = readonly ? SymbolFlags.READONLY : 0;
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java:67]: Can we use this(memberTypes, memberTypes,typeFlags, isCyclic, readonly ? SymbolFlags.READONLY : 0)

**Comment by warunalakshitha** [comment]
LGTM

**Review by grainier — APPROVED** [review]
LGTM

**Review by grainier — APPROVED** [review]

**Comment by MaryamZi** [comment]
> @MaryamZi There is a local build failure with this PR with parallel merge.
> 
> failing following, `./gradlew :ballerina-runtime:test`
> 
> output
> 
> ```
> ballerina-runtime-test-suite > ballerina-runtime-test > io.ballerina.runtime.test.cli.OptionTest > testInvalidUnionType FAILED
>     java.lang.AssertionError: expected [unsupported type expected with main function 'string|int'] but found [unsupported type expected with main function '(string|int)']
>         at org.testng.Assert.fail(Assert.java:96)
>         at org.testng.Assert.failNotEquals(Assert.java:776)
>         at org.testng.Assert.assertEqualsImpl(Assert.java:137)
>         at org.testng.Assert.assertEquals(Assert.java:118)
>         at org.testng.Assert.assertEquals(Assert.java:453)
>         at org.testng.Assert.assertEquals(Assert.java:463)
>         at io.ballerina.runtime.test.cli.OptionTest.testInvalidUnionType(OptionTest.java:145)
> ```

Fixed in 5ba80b5, thanks.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java (hunk 3 above)
Comment (MaryamZi) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java]: Fixed in 484ec9f, thanks!

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java (hunk 2 above)
Comment (MaryamZi) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java]: Yeah, updated in a422002, thanks!

**Review by grainier — APPROVED** [review]

**Pull request merged**